Game refactor for testability

lphillips opened the pull request:
This is very much a work-in-progress of a refactor of the game logic to improve testability. The aim of this refactor is to:

- Move state management into the game module
- Reduce code duplication
- Isolate hard to test code to allow testing of the code around it

I've focused on the code around voting so far, since that is one of the main inputs that drives state changes currently. This means updating the `Game.vote(player_id, vote)` method and `on_reaction_add(reaction, user)` function, with some changes radiating out from there.

In order to centralize state transitions and have a common flow that is executed when state changes, I've introduced `Game.__transition_to_state(state)` to isolate assignments to `Game.state` and also add methods that can tell an observer when the game has entered a new state. The idea here is that the app module sends inputs to the `Game` via methods like `vote`, `execute_player`, `nominate`, etc, and that can trigger state changes. When the state has changed, the app is notified via a method like `on_legislative_presidential`, so it can created the appropriate embed and send it to Discord. 

I've run into one hiccup here though. Messaging with Discord is asynchronous, so I can't call from the synchronous state management code in `Game` to the app module to update the UI. I'm very new to async programming in Python, but I'm currently thinking of setting up some async dispatch to the embed creation in response to the game callback. Something like this:

```
# In app.py
def on_legislative_presidential(game):
   # dispatch_to_async() is a stand-in for whatever asyncio method I can call to get it to run
   # an async method and not wait for the result. Also need to send the game through to
   # async method, however that works
    asyncio.dispatch_to_async(send_embed_for_legislative_presidential)

async def send_embed_for_legislative_presidential(game):
    # use discord API to create and send an embed
    game.finish_legislative_presidential() # this call triggers the game to move to the next state
```

lphillips committed:
refactor WIP

bsmccain commented:
I think asyncio.create_task() is what you want to make the async call.  That's what I did in the emit() method on DiscordChannelHandler.
Still reviewing, wanted to make this comment before I forgot.

lphillips commented:
`asyncio.create_task()` seems to do the trick. I set up the code for creating a game and reacting to the game_start state change and everything happened as expected.

bsmccain reviewed (approved):
I like the direction this is going, and approve the current implementation.  I think if I were doing this from scratch, I'd probably create state objects that had methods for handling voting, transitions, etc.  I'm not certain it's worth the effort in this case, but the large if/else trees seem unwieldy.  But they're also implemented and functioning currently, so just rearranging is probably fine.

lphillips committed:
Add GameStateObserver

Additional refactoring for testability.

This commit adds a GameStateObserver so the game can message back to the app about state changes. More testing is added as well.

lphillips committed:
Additional refactoring and testing

lphillips commented:
I think this is in a state where it is ready for further review and possible merging. Here's a summary of the current state:

- `game.py` has been refactored to achieve the goals of reducing code duplication and making it more easily testable.
- `test_game.py` has an extensive suite of tests that covers a lot of the functionality in the game
- A `GameStateObserver` has been added to message back to the app about state transitions in the game, so the app doesn't have to make assumptions about when the state will change. There's a type-insensitive "context" parameter that gets passed through the system, so the app can pass the Discord context into the game and when the state changes, the game passes that context back via the `GameStateObserver` so we keep our guild context intact for sending messages.
- Image handling was factored out to a new class in order to make testing easier and to remove the duplicate code for handling images. There's no testing for this new class currently.

And a caveat:

- `app.py` is only partially integrated with these changes currently, so the bot won't really work with the current state of this PR. This is big enough that I'm kind of inclined to get this reviewed and merged and do a follow-up refactoring PR for the app. Open to suggestions on this though.

lphillips commented:
> I like the direction this is going, and approve the current implementation. I think if I were doing this from scratch, I'd probably create state objects that had methods for handling voting, transitions, etc. I'm not certain it's worth the effort in this case, but the large if/else trees seem unwieldy. But they're also implemented and functioning currently, so just rearranging is probably fine.

Yeah, I considered pulling in a state machine library and building an entirely new game engine off of that, but ultimately decided I could achieve good results by moving some of this code around and reducing duplication. A lot of the deeply nested if-statements are gone now, so it is easier to read and reason about.

lphillips committed:
Merge branch 'develop' into refactor-command-handling

bsmccain commented on secret_hitler/app.py:
```diff
@@ -103,105 +154,15 @@ async def on_reaction_add(reaction, user):
         return
 
-    if game.state == GameStates.ELECTION:
+    if game.state == GameState.ELECTION:
```
I wouldn't check game state here.  This should convert the vote, send it to the Game, and let the Game determine if voting is allowed or not.

bsmccain replied:
Looks like game state is still being checked all over the place in this file. Not going to comment on each place, but that logic should all get moved into Game.

bsmccain commented on secret_hitler/game.py:
```diff
@@ -49,55 +78,177 @@ def __init__(self, channel_id, game_id, max_players, admin_id):
+        if self.get_player(player_id) is None:
+            return False
+
+        for pid, s_v in self.votes.items():
```
can't this just be 
```
if player_id in self.votes:
    return False
```

lphillips replied:
Yeah, I'm pretty sure that is correct and is more concise, so I'll make that change.

bsmccain commented on secret_hitler/game.py:
```diff
@@ -49,55 +78,177 @@ def __init__(self, channel_id, game_id, max_players, admin_id):
+            if pid == player_id:
                 return False
 
         if len(self.votes) == len(self.players):
```
Can this case happen?  If the player is in the game (got past first if), and they haven't voted (got past second if), then the votes length and players length can't be equal because this player hasn't voted.  Not a huge deal, but I don't believe it can ever be true.

If it were placed above the previous if, or even at the beginning of the function, it would reduce the need for additional searching.   But with lists this small, it's not going to make a difference

lphillips replied:
Yeah, this case seems redundant.

bsmccain commented on secret_hitler/app.py:
```diff
@@ -31,6 +31,57 @@ def emit(self, record):
                 self.handleError(record)
 
 
+class GameStateObserver:
```
This design is a little weird to me.  Maybe it's just a python thing hurting my java brain, but I'd expect Game to define the interface for the callback functions.  The app would subclass it and then implement those functions, then pass the implemented observer to the Game when it starts.  You'd still have this class in app.py, I guess, with an interface of the same functions in game.py, so maybe that's not python-esque.

bsmccain replied:
With that said, there's a lot of redundancy here, and my brain is going straight to reflection to simplify this class.  It's really just translating the callbacks to functions defined here, so you could do something like
```
class GameStateObserver:
    def __getattr__(self, function):
        def f(*args, **kwargs):
            if function in globals():
                return globals()[function](*args, **kwargs)
            else:
                return handle_unknown_callback(*args, **kwargs)
        return f
```
and then just define all the on_enter* functions once in app.py.  I guess you could do something similar without using reflection by just making everything in app a class, then passing that object to the Game, then it could call the methods directly there.  Regardless, it still feels a little weird to me that the definition of the callbacks are not part of Game.

lphillips replied:
I looked a bit into something like Java's `interface`s in Python, and pretty recently in 3.8 they added "protocols" which is the same basic idea. However, Python is mostly built around the idea that the programmer knows what they are doing and will make sure interfaces match up. I would have implemented these methods on an App class...but that doesn't exist. The app is a ball of global methods. So, I made a little class to implement these methods and map to the app's messaging methods with an appropriate asynchronous wrapper.

It's not especially structured with types and interfaces, so....pretty Python-y? Your proposed change is even more Python-y though so I'll probably give that a try.

bsmccain commented on secret_hitler/app.py:
```diff
@@ -1226,6 +1178,125 @@ async def print_help(channel):
+
+
+# ============================================
+# App start up
```
I don't think this belongs here

lphillips replied:
True. Will move.

bsmccain commented on secret_hitler/game.py:
```diff
@@ -49,55 +78,177 @@ def __init__(self, channel_id, game_id, max_players, admin_id):
+            vote_passed = self.__calculate_votes()
+            next_state = self.__next_state_after_vote(vote_passed)
+            self.__transition_to_state(next_state)
+            if vote_passed:
```
I don't think we need to return additional VoteResult types here...the call is when someone clicks an emoji, either that click is valid or not.  Yes, it is also going to tally the votes once they are all in, but that info will get reported to the UI (app) via the callback mechanism during the state transition.  The caller has no reason to care what the outcome of the vote is in this context, it only cares if its vote was or was not valid.

lphillips replied:
I wrote the VoteResult enum early on in the refactor and did not revisit it, but yeah, at this point, all that matters is indication of if the vote was counted or not.

bsmccain reviewed (commented):
A few comments about design points and implementation in places.  Would like your thoughts on the comments before merging.  However, a lot of them could be done separately, since this PR breaks things anyway, maybe it will be easier to merge it and move on with additional changes at this point.

lphillips committed:
Address some PR comments

lphillips commented:
I addressed the smaller items from your comments. Here's what I plan to do in my next PR:

- Update GameStateObserver per your comments
- Refactor `app.py` to no longer reference Game.state. We'll rely on exception handling instead.
- Update `VoteResult` and the `vote` method to use exceptions and return a value that indicates if the vote was counted. Might just switch it to a bool if that makes sense.
